internal/pkg/service/common/etcdop/op/atomic.go

```diff
@@ -291,35 +292,35 @@ func (v *AtomicOp[R]) DoWithoutRetry(ctx context.Context, opts ...Option) (bool,
 		}
 	}
 
-	// Create transaction
-	txn, err := v.createWriteTxn(ctx)
+	// Create WRITE transaction
+	writeTxn, err := v.createWriteTxn(ctx)
 	if err != nil {
 		return false, err
 	}
 
 	// Add IF conditions
-	txn.If(cmps...)
+	writeTxn.If(cmps...)
 
 	// Do
-	txnResp := txn.Do(ctx)
-	return txnResp.Succeeded(), txnResp.Err()
+	writeResult := writeTxn.Do(ctx)
+	return writeResult.Succeeded(), writeResult.Err()
 }
 
 func (v *AtomicOp[R]) createWriteTxn(ctx context.Context) (*TxnOp[R], error) {
 	// Create WRITE operation
-	var writeOps []Op
+	writeTxn := TxnWithResult[R](v.client, v.result)
 	for _, opFactory := range v.writePhase {
 		op, err := opFactory(ctx)
 		if err != nil {
 			return nil, err
 		}
 		if op != nil {
-			writeOps = append(writeOps, op)
+			writeTxn.Merge(op)
 		}
 	}
 
 	// Create transaction
-	return TxnWithResult[R](v.client, v.result).Then(writeOps...).AddProcessor(func(ctx context.Context, r *TxnResult[R]) {
+	return writeTxn.AddProcessor(func(ctx context.Context, r *TxnResult[R]) {
```

Comment:
Write phase: `Merge` instead of `Then` bellow, and removed unnecessery aux slice.